orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+                    raise RuntimeError("Unexpected keys: {} in state_dict[optimizer][{}]"\
+                        .format(unexpected_keys, model_state_key))
+                missing_keys = list(training_session_optimizer_state_keys - state_dict_optimizer_state_keys)
+                if len(missing_keys):
+                    raise RuntimeError("Missing keys: {} in state_dict[optimizer][{}]"\
+                        .format(missing_keys, model_state_key))
+
+        def _check_key_mismatch(training_session_state_dict, state_dict):
+            model = 'model'
+            optimizer = 'optimizer'
+            # check presence of 'model' in the input state_dict
+            if model not in state_dict:
+                raise RuntimeError("Missing key: model in state_dict")
+            # check presence of 'optimizer' in the input state_dict
+            if optimizer not in state_dict:
+                raise RuntimeError("Missing key: optimizer in state_dict")
+
+            _check_model_key_mismatch(training_session_state_dict, state_dict)
+
+            _check_optimizer_key_mismatch(training_session_state_dict, state_dict)
+
+
+        # extract state dict from the current training session. this is to persist the states between
+        # two training sessions.
+        # for example, if user provided only the model states, the optimizer states from the current
+        # training session must be persisted
+        states = {}
+        if self._training_session:
+            states = self.state_dict()
+            if strict:
+                _check_key_mismatch(states, state_dict)
+
+        # load the model states from the input state dictionary into the onnx graph and update the
+        # training session states dictionary
+        self._load_model_states(states, state_dict, strict)
+
+        # load the optimizer states from the input state dictionary into the training session states
+        # dictionary
+        self._load_optimizer_states(states, state_dict)
+
+        return states
+
+    def load_state_dict(self, state_dict, strict=True):
+        """Loads input state dictionary containing model and optimizer states into the training session.
+
+        - Load the model states from input state dictionary
+        - Load the optimizer states from input state dictionary
+        - If onnx graph has not been initialized, the states will not be immediately loaded onto the graph;
+        the states will be loaded once the training session has been initialized when either train_step or eval_step is called
+
+        Args:
+            state_dict: state dictionary containing both model and optimizer states. The structure of this dictionary
+                should be the same as the one that is returned by ORTTrainer.state_dict for the case when pytorch_format=False
+            strict: boolean flag to strictly enforce that the input state_dict keys match the keys from ORTTrainer.state_dict
+        """
```

Comment:
```suggestion
        """Loads state_dict containing model/optimizer states into ORTTrainer

        The state_dict dictionary may contain the following information:
           - Model and optimizer states
           - Required ORTTrainerOptions settings
           - Distributed training information, such as but not limited to ZeRO

        Args:
            state_dict: state dictionary containing both model and optimizer states. The structure of this dictionary
                should be the same as the one that is returned by ORTTrainer.state_dict for the case when pytorch_format=False
            strict: boolean flag to strictly enforce that the input state_dict keys match the keys from ORTTrainer.state_dict
        """
```

Users shouldn't know about any ONNX graph initialization, so we either use a neutral terminology (without implementation detail) or we suppress if there is nothing the user can do about it